python/django/security/injection/open-redirect.yaml

```diff
@@ -15,517 +16,37 @@ rules:
     - django
   languages: [python]
   severity: WARNING
-  patterns:
-  - pattern-inside: |
-      def $FUNC(...):
-        ...
-  - pattern-not-inside: |
-      def $FUNC(...):
-        ...
-        django.utils.http.is_safe_url(...)
-        ...
-  - pattern-not-inside: |
-      def $FUNC(...):
-        ...
-        if <... django.utils.http.is_safe_url(...) ...>:
+  pattern-sources:
+  - patterns:
+    - pattern-inside: |
+        def $FUNC(...):
           ...
+    - pattern-not-inside: |
+        def $FUNC(...):
+          ...
+          <... django.utils.http.is_safe_url(...) ...>
+          ...
+    - pattern-not-inside: |
+        def $FUNC(...):
+          ...
+          <... django.utils.http.url_has_allowed_host_and_scheme(...) ...>
+          ...
+    - pattern-not-inside: |
+        def $FUNC(...):
+          ...
+          if <... django.utils.http.is_safe_url(...) ...>:
+            ...
+    - pattern-not-inside: |
+        def $FUNC(...):
+          ...
+          if <... django.utils.http.url_has_allowed_host_and_scheme(...) ...>:
+            ...
+    - pattern-not: request.get_full_path(...)
+    - pattern-either:
+      - pattern: request.$W.get(...)
```

Comment:
You forgot:
```
- pattern: request.$W
```

Reply:
Actually if we add `request.$W` as in the search-mode rule, I guess we don't need `request.$W.get(...)`.